### PR TITLE
pagecache benchmarks

### PR DESCRIPTION
**Pull request by achille-roussel**
I was curious to see the impact of messing with some of the parameters like the bucket count or the size of reads, so I added a few benchmarks.

It seems that there was still quite a bit of lock contention when increasing the number of threads in the test, and increasing the bucket count effectively reduces it and improves overall throughput of the datastructure.

Here are the results from the synthetic benchmarks highlighting the effect of increasing the bucket count from 64 to 512:

```
name                       old time/op  new time/op  delta
PageCacheNoEvictions-12    67.9ns ± 3%  34.9ns ± 2%   -48.61%  (p=0.000 n=10+10)

name                       old hits(%)  new hits(%)  delta
PageCacheNoEvictions-12       100 ± 0%     100 ± 0%      ~     (all equal)

name                       old read/s   new read/s   delta
PageCacheNoEvictions-12     17.5M ± 3%   34.1M ± 2%   +94.60%  (p=0.000 n=10+10)
```

Beyond 512, throughput started decreasing; it is hard to know for sure what dynamics are at play, it may have to do with having the cache datastructure exceed the CPU cache size or hitting similar limits.

I also moved the `[]byte` containing the page data from having a subslice referenced in each bucket to having it tracked by the `Cache` instead, which reduces the memory footprint of the cache a bit.

I also testing moving the stats counters to the `Cache` but this actually had a significant negative impact on throughput; with a mutex we observe higher contention due to having a global lock to update the counters, and with atomic operations I also saw an even worse impact, likely due to atomic instructions having significantly higher latency.

I tested all these on a MacBook M1, different CPUs may behave slightly differently.

### CODE REVIEW

File: pagecache/pagecache.go [hunk 1]
```diff
@@ -35,7 +35,7 @@ const (
 	// ensure that we retain the same performance characteristics, which would
 	// require us to only allow powers of two as bucket counts, and implement
 	// the bitwise optimizations in the code.
-	numBuckets = 64
+	numBuckets = 512
```
Comment (kevinburkesegment) [inline pagecache/pagecache.go:41]: Put a link here to the investigation? If you did the work might as well.

File: pagecache/pagecache.go [hunk 2]
```diff
@@ -132,19 +133,24 @@ func NewWithConfig(config *Config) *Cache {
...
+	for i := range pages {
+		pages[i].offset = uint32(i)
+	}
+
+	bucketSize := len(pages) / len(c.buckets)
 	for i := range c.buckets {
-		b := &c.buckets[i]
-		b.init(data[int64(i)*bucketSize:int64(i+1)*bucketSize], pageSize)
+		off := (i + 0) * bucketSize
+		end := (i + 1) * bucketSize
```
Comment (kevinburkesegment) [inline pagecache/pagecache.go:155]: Thanks, this is much more readable.

File: pagecache/pagecache.go [hunk 3]
```diff
@@ -190,6 +202,12 @@ type Stats struct {
 	Frees     int64 // number of allocated pages returned to the free pool
 }
 
+// HitRate returns the hit rate of cache lookups, as a floating point value
+// between 0 and 1 (inclusive).
+func (s *Stats) HitRate() float64 {
+	return float64(s.Hits) / float64(s.Lookups)
```
Comment (kevinburkesegment) [inline pagecache/pagecache.go:214]: Divide by zero error?

**Review by kevinburkesegment — APPROVED** [review]
